tools/README.md

```diff
@@ -0,0 +1,72 @@
+
+```json
+{
+	"scripts": {
+		"build": "pnpm --if-present --workspace-concurrency=Infinity --stream --filter=\"$npm_package_name...\" '/^build:project:.*$/'",
+    "build:project": "pnpm --if-present '/^build:project:.*$/'",
+		"lint": "pnpm --if-present '/^lint:lang:.*$/'",
+		"lint:fix": "pnpm --if-present '/^lint:fix:lang:.*$/'",
+		"watch:build": "pnpm --if-present --workspace-concurrency=Infinity --filter=\"$npm_package_name...\" --parallel '/^watch:build:project:.*$/'",
+    "watch:build:project": "pnpm --if-present run '/^watch:build:project:.*$/'"
+	}
+}
+```
+
+These scripts outline the naming scheme used in order to facilitate [task parallelization using regular expressions](https://pnpm.io/cli/run#running-multiple-scripts). **To ensure consistency across the monorepo, these scripts should not be edited.** New scripts should be added using the naming scheme outlined by the above regular expressions, for example, `build:project:bundle` might be a script to run a tool like `webpack`. We also utilize a number of PNPM options to ensure a positive developer experience:
+
+- `--if-present`: Ensures that PNPM will not error if a script is not found.
+- `--workspace-concurrency=Infinity`: Runs as many of the tasks in parallel as possible.
+- `--stream`: Makes the script output legible by putting all of their output into a single stream.
+- `--filter="$npm_package_name..."`: This filter tells PNPM that we want to run the script against the current project _and_ all of its dependencies down the graph (dependencies first).
+
+To further improve the build times, we used two additional techniques:
+
+- In the case of the `build` script, we offer both building packages with (`build`) and without (`build:project`) its dependencies.
+- Using `wireit`-based task output caching (details are below).
+
+## Task Output Caching
+
+Our repository uses [`wireit`](https://github.com/google/wireit) to provide task output caching. In particular, this allows us to cache the output of `build` scripts so that they don't run unnecessarily.
+The goal is to minimize the amount of time that developers spend waiting for projects to build.
+
+```json
+{
+  "scripts": {
+    "build": "pnpm --if-present --workspace-concurrency=Infinity --stream --filter=\"$npm_package_name...\" '/^build:project:.*$/'",
+    "build:project": "pnpm --if-present '/^build:project:.*$/'",
+    "build:project:bundle": "wireit"
+  },
+  "wireit": {
+    "build:project:bundle": {
+      "command": "webpack",
+      "files": [
+        "... - package resources as input"
+      ],
+      "output": [
+        "... - package resources as output"
+      ],
+      "dependencies": [
+        "dependencyOutputs"
+      ]
+    },
+    "dependencyOutputs": {
+      "files": [
+        "... - dependencies resources as input"
```

Comment:
This is actually[ a PNPM hook in the `.pnpmfile.cjs` at the monorepo root](https://github.com/woocommerce/woocommerce/blob/7d5d032145f67114f824a9f83b57c7dd2e787438/.pnpmfile.cjs). There's some [documentation on how these hook files work](https://pnpm.io/pnpmfile). Essentially, when our dependencies are updated, it will generate the list again. This works because the only reason these dependencies would change is if the dependencies on the package changed and that would trigger this script.

(I don't have GitHub notifications silenced because I maintain my VS Code extension :smile:)

Reply:
Here we go: e570408b8528bf6944e07abe6fb54c736bb2c69f, b701f314c7adb63912f47fb240cafe0fc949e7f5